added t-digest structure

catpouch opened the pull request:
it's in the title
the following article was most helpful for me in understanding t-digest (but there's some stuff missing/different/wrong):
https://www.gresearch.com/blog/article/approximate-percentiles-with-t-digests/
also refer to the original paper:
https://arxiv.org/abs/1902.04023

i left a handful of comments to explain some things that aren't explained in either of those but idk how helpful they are

catpouch committed:
added t-digest structure

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+    //     digest
+    // }
+
+    // apparently, this is not a std function
```
The `num` crate has a lot of stuff like this and is so commonly used it's basically an extension of the std lib. https://docs.rs/num/latest/num/fn.clamp.html

catpouch replied:
still feels nicer to me to just write this function rather than drag in an entire dependency to only use one part of it
i'll get rid of the comment though

Psykopear replied:
This is perfectly fine, but if you don't like it, another way you could write it is `v.max(lo).min(hi)`. Also, looking at the source code of the implementation in `num` [here](https://docs.rs/num-traits/0.2.15/src/num_traits/lib.rs.html#408), it's probably good to add an assert to check that `lo <= hi`:
```rust
fn clamp(v: f32, lo: f32, hi: f32) -> f32 {
    debug_assert!(lo <= hi, "lo must be less than or equal to hi");
    ...
}
```

catpouch replied:
added this suggestion. the function was only there because i copied it (along with `estimate_value`) from a different crate after realizing that if i wrote `estimate_value` myself it would end up being fundamentally the same code.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+pub struct TDigest {
+    centroids: Vec<Centroid>,
+    delta: f32,
+    buffer: [f32; 32], // used exclusively for add_value
```
A `Vec` alone could manage this buffer and index for you. `Vec::with_capacity(32)`, then `buffer.push`, then `if buffer.len() == buffer.capacity() { buffer.drain() }`.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+    pub fn add_value(&mut self, value: f32) {
+        self.buffer[self.index as usize] = value;
+        self.index += 1;
+        if self.index == 32 {
```
The buffer is not drained before calculating any statistics so they are not "correct". For large datasets, skipping the most recent 31 values might not matter, but for small ones the algorithm should still work and would return incorrect values.

catpouch replied:
this is intentional, it's supposed to be used with the knowledge that the structure won't be completely up-to-date at any given time but individual merges will be much faster. making it so that the buffer is drained before any reads would mostly defeat the purpose, since you're just relocating the latency that comes with merging values (it's the slowest operation by far).
i did plan on adding more descriptive docstrings to make this clear later on

Psykopear replied:
You could expose a method that drains the internal buffer manually though, so that if a user needs exact calculations, they can call the method before reading statistics

davidselassie replied:
If you want to assume that the hot path is adding data and needing the statistics is a less frequent event, you could call the "flush buffer" method whenever the statistics are requested to ensure that they are correct. Then a separate public method isn't needed.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+
+    /// Adds a single value into the T-Digest, using amortization.
+    /// Does batch updates to the T-Digest every 32 added values. In the meantime, the structure will not change.
+    pub fn add_value(&mut self, value: f32) {
```
I think the buffer management fits best one level up: Don't have a `TDigest::add_value` and only have `TDigest::merge_vec`, then you could isolate the buffer in the Python layer and call `inner.merge_vec` when appropriate. The pyclass would still have an `add_value` and would use the buffer.

catpouch replied:
i wrote this as a rust crate first and then just copied everything over with as few changes as possible, so that's why this is a bit clunky. i'll consider fixing it, but at this point it's such a simple thing to do that it might be best to remove it as a feature and leave it up to the user.
i originally added it because the other rust crate that implements t-digest had an unmerged PR that had a second variation to add this feature

davidselassie commented on src/lib.rs:
```diff
@@ -284,12 +287,44 @@ impl CountMinSketch {
+
+    fn merge_vec_unsorted(&mut self, v: Vec<f32>) {
+        self.inner.merge_vec(v);
+    }
```
Maybe name these `push` and `push_list` to match the API of the other classes in this lib?

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+    // not sure if this function could be optimized, but its usage can
+    // gets the weight-based quantile of a centroid in the digest (given by index)
+    fn centroid_quantile(&self, index: usize) -> f32 {
+        let weight_to_index = self.centroids.iter().take(index + 1).fold(0, |sum, x| sum + x.weight) as f32;
```
Is there a reason for some of these you use `fold(0, |sum, x| sum + x.weight)` and on some you use `map(|c| c.weight).sum()`? I find the latter much more readable.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+    // binary search to find the centroid with the next largest mean after given mean
+    // will typically use O(log n) time, could take less in the case of an exact match
+    // not sure how much optimization there is to be had here aside from a better algorithm
+    fn search_centroids(&self, mean: f32) -> usize {
```
I believe you could use [`Vec::binary_search_by`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by) or one of the related `binary_search` methods instead of re-implementing this.

catpouch replied:
looked into it, none of them would work since i'm doing a search by minimum distance and not an exact search

davidselassie replied:
Isn't that the point of the map function `f` supplied to the `_by` version? You can have that function calculate the distance in mean then use the `Err` result containing the index of what would be the cluster next largest mean.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+}
+
+impl Centroid {
+    fn new(mean: f32, weight: i32) -> Self {
```
FYI there's no requirement to have a `new` function on a private struct. I'd construct the struct directly via `Centroid { .. }` elsewhere in this file and only add a `new` function if there's some sort of invariant or setup that you don't want to repeat.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+
+/// T-Digest algorithm. A probabilistic data structure for estimating quantiles of data while using a fixed amount of memory.
+pub struct TDigest {
+    centroids: Vec<Centroid>,
```
One thing that would be fun to benchmark here would be to transform this into a struct-of-arrays (rather than the array-of-structs that is currently is). Since most of your iteration or searching is solely across either means or weights, this might provide a significant speed up. It would make sorting centroids far more complicated, though, and you'd probably need to pack into tuples and then unpack back into the arrays. I don't think you need to do this, but if you're interested in another experiment, could be fun.

catpouch replied:
it's definitely an interesting idea, i may look into it in the future but for now i'd rather leave the codebase as is. maybe i'll revisit it during one of my longer school year breaks.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+    
+    // wanted to add a sorted version of this but i couldn't think of a version to save that much time and this is pretty fast already. if you wanted to make it super duper fast that's what you would do i guess
+    /// Merges an unsorted vector of values into the T-Digest.
+    pub fn merge_vec(&mut self, v: Vec<f32>) {
+        self.centroids.extend(Self::vals_to_centroids(v));
```
As is, there are 2 extra allocations to add values, `self.buffer.to_vec()` and then `Self::vals_to_centroids` does `collect` within. You could have `merge_vec` take `v: impl IntoIterator<Item = f32>` and then `merge_vec(buffer.drain())` would work without any allocations.

catpouch replied:
true, but this function is supposed to be public, and i feel that it would be more awkward to ask the user to pass in an `IntoIterator` than have to do a quick allocation before merging (especially since the idea behind the buffer is an occasional expensive draining while being fast most of the time).
i will see if i can figure something out for the `vals_to_centroids` thing though

Psykopear replied:
I agree with david, it's usually better to allow any kind of iterator if it doesn't cost anything, and in this case you also gain something.

From the public api point of view, users can still pass a `Vec` if they have it already, but if they don't you allow them to pass an iterator too.

This function can accept an `impl IntoIterator<Item = f32>`, so can `Self::vals_to_centroid` (you change from `.iter` to `.into_iter`), and then you can change the return type of `vals_to_centroid` to `impl IntoIterator<Item = Centroid>`, and avoid the `collect` since the `extend` here accepts an iterator too.

catpouch replied:
oops, didn't realize that `impl IntoIterator` accepted vectors. rewrote things to accept this suggestion

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+        let mut k_limit: f32 = 1.0; // if you read the gresearch article, it's actually slightly wrong. the new group of centroids doesn't start at the one that pushes the delta k past 1, it starts at each k whole number (so that you don't end up with >delta centroids)
+        let mut q_lower: f32 = 0.0; // lower quantile limit. feel like there's a small optimization to be had here but i'm not seeing it.
+        let mut q_limit = inverse_scale(k_limit, self.delta); // precomputed as a max quantile value here & after each new centroid is completed so that you don't have to compute new k values with every single iteration
+        let mut current_centroid = self.centroids[0]; // also used as a buffer, stores the aggregated centroid that is to be added to temp_centroids
+        for i in 1..self.centroids.len() {
```
Is there a better "zero value" here for `current_centroid`? Maybe something like `Centroid { 0, 0 }`? Then you could iterate through the centroid without needing to slice off the first element and could do `for centroid in &self.centroids`.

catpouch replied:
that would be cleaner, but i feel like this is not super unreadable and it saves on One Whole Iteration. unless vector accesses are really slow and i just didn't know that, in which case let me know.

Psykopear replied:
Another way to write this and avoid the initial slicing is using the `fold` iterator. For example something like this, it doesn't use `self.temp_centroids`, and does allocate a new vec each time the function is called, but maybe something you can build upon:
```rust
        self.centroids = self.centroids.iter().fold(Vec::new(), |mut acc, centroid| {
            // In the first iteration, just add the centroid to the acc
            let Some(prev_centroid) = acc.last_mut() else {
                acc.push(*centroid);
                return acc;
            };
            // Then either merge the current and previous centroids, or update the limits
            // and push the centroid as it is.
            let q_upper =
                q_lower + (prev_centroid.weight + centroid.weight) as f32 / weight_sum as f32;
            if q_upper <= q_limit {
                prev_centroid.merge(centroid);
            } else {
                q_lower += prev_centroid.weight as f32 / weight_sum as f32;
                k_limit += 1.0;
                q_limit = inverse_scale(k_limit, self.delta);
                acc.push(*centroid);
            }
            acc
        });

```

catpouch replied:
while it does look nicer, even if i found a way to avoid the vector allocation i can't see it providing a worthwhile performance benefit. plus (at least to me) a for loop is a bit easier to wrap my head around. since the initial slice isn't actually manipulating any vector elements, i can't imagine it taking up any significant portion of the function's runtime

Psykopear replied:
The performance benefits can also come from things you don't immediately notice in the source code, depending on how the code is compiled and optimized ([an interesting article on the argument](https://matklad.github.io/2023/04/09/can-you-trust-a-compiler-to-optimize-your-code.html)). Now, I'm not that good at understanding the generated assembly, but I tried to compare the two different version on godbolt [here](https://godbolt.org/z/9vYbEGE8z), and the one with the iterator seems to do the same thing with a lot less code.
This doesn't necessarily mean anything, but it shows that the 2 implementation end up being quite different. This usually makes me want to setup a benchmark that stresses this specific loop to see if there's an actual difference.

Not saying you need to do any of that, it's already impressive what you achieved in just a few weeks here, I just wanted to share my thinking process in such cases

davidselassie replied:
I agree with Fede's suggestion of using `fold` if possible. I try to avoid index (or pointer) twiddling as much as possible because, although it is "easier" to code, it is also easier to make small mistakes in indexing and iteration, and unless there's a giant performance detriment (of which Fede also points out, there rarely is), I appreciate the stronger guarantee of correctness. Having to use a well-known iteration pattern (via `fold` or `map` or `sum`) constrains the code space so much in a good way that it makes it harder to make correctness mistakes. Especially typos. I'd encourage you to read up on functional programming patterns for the kinds of shapes available.

davidselassie commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+    /// Estimates what quantile a given value is in.
+    pub fn estimate_quantile(&self, v: f32) -> f32 {
+        let index = self.search_centroids(v);
+        if index == 0 {
+            return 0.0;
+        } else if index == self.centroids.len() - 1 {
+            return 1.0;
+        }
```
Should this result in a LERP between 0 and the mean of the index for the `index == 0` case and between the mean of the index and 1 for the `index == len - 1` case? It seems like there should be a way to provide more accuracy here?

catpouch replied:
`index == 0` means that the very first centroid has been selected, which will always represent the beginning of the dataset, so the quantile there is 0. as for the other case, you would be correct but the reason is that i actually just wrote that code wrong. it should trigger only when you've selected a value above the upper bound of the dataset, so i changed some things about the search algorithm to make that the case.

davidselassie replied:
This implementation still seems to be ignoring this paragraph from the paper:

> The final case to be considered is that of the first or last cluster in the digest. When
such a terminal cluster has only one sample, then we know that xmax is due exactly to that
single sample. Where the cluster has two samples, we know that those two samples were
at xmax and xn − (xmax − x) and thus they can be treated as two singletons. Where there
are more than two samples, however, it is advantageous to treat the cluster as having half
the samples before the centroid and half after. For the half of the samples that come after, one of those samples must be at xmax and thus it can be treated as a singleton with the
remainder being interpolated. All of this discussion is reversed, of course, first for last and
left for right when treating the first cluster. This is shown in Figure 5 for the last cluster
in a digest.

That is implying it is possible to introspect the first or last cluster to produce a better estimate beyond the means of that cluster.

davidselassie reviewed (commented):
Good work on this. Writing out nitty-gritty algorithmic code is hard. Lots of little comments here.

Psykopear commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+
+impl Centroid {
+    fn new(mean: f32, weight: i32) -> Self {
+        Self {
+            mean: mean,
+            weight: weight,
+        }
```
Rust tip: whenever the variable has the same name of the struct field, you can use the equivalent shorthand initialization:
```rust
Self { mean, weight }
```

Psykopear commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+
+    // does what it says. probably not necessary since it's only used once
+    fn vals_to_centroids(v: Vec<f32>) -> Vec<Centroid> {
+        v.iter().map(|x| Centroid::new(*x, 1)).collect()
```
Since you only use `Centroid::new` here, maybe you can modify it to use `1` as default value for weight:
```rust
pub fn new(mean: f32) -> Self {
    Self { mean, weight: 1 }
}
```

So that here you can write:

```rust
v.iter().map(Centroid::new).collect()
```

Psykopear commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+        // the following chunk of code does the same thing as search_centroids but it's based on quantile instead of mean and also it's much faster for the use case (most queries are going to be at the extremes, so linear search works out to be pretty fast)
+        let mut pos: usize;
+        let mut t: f32;
+        if q > 0.5 {
+            if q >= 1.0 {
+                return self_max;
+            }
+
+            pos = 0;
+            t = count_;
+
+            for (k, centroid) in self.centroids.iter().enumerate().rev() {
+                t -= centroid.weight as f32;
+
+                if rank >= t {
+                    pos = k;
+                    break;
+                }
+            }
+        } else {
+            if q <= 0.0 {
+                return self_min;
+            }
+
+            pos = self.centroids.len() - 1;
+            t = 0.0;
+
+            for (k, centroid) in self.centroids.iter().enumerate() {
+                if rank < t + centroid.weight as f32 {
+                    pos = k;
+                    break;
+                }
+
+                t += centroid.weight as f32;
+            }
+        }
```
You could use a `match` statement here:

```rust
        match q {
            q if q <= 0.0 => return self_min,
            q if q >= 1.0 => return self_max,
            q if q > 0.5 => {
                ...
            }
            _ => {
                ...
            }
        };
```

Psykopear commented on src/tdigest.rs:
```diff
@@ -0,0 +1,269 @@
+        }
+
+        let value = self.centroids[pos].mean + ((rank - t) / self.centroids[pos].weight as f32 - 0.5) * delta;
+        Self::clamp(value, min, max)
```
Since you are only using this once here, as suggested above, you could write `value.max(min).min(max)` :smiling_face_with_tear:

catpouch committed:
lots of small changes
-removed buffer
-cleaned up a bunch of code
-fixed slightly broken functionality in estimate_quantile
-split merge_vec into sorted & unsorted
-haven't added tests yet so this all might be completely broken

catpouch committed:
slight tdigest change + tests added / updated

davidselassie commented on src/tdigest.rs:
```diff
@@ -244,26 +240,19 @@ impl TDigest {
-            self.index = 0;
-        }
+    /// Merges a sorted vector of values into the T-Digest.
+    pub fn merge_vec_sorted(&mut self, v: impl IntoIterator<Item = f32>) {
```
Although there's a balance to be struck between performance and API ergonomics, I generally don't think exposing an "insert already sorted data" API is the right call. Optimized sorts are pretty fast and unless you have measured a big performance win _and_ there's a large known use case that involves having pre-sorted data ready to be used here, I am always reticent to introduce and API that assumes anything about the data. Especially one that fails silently; AFAICT if you insert unsorted data, you'll get wrong answers sometimes, not panics.

I would drop this public API, or at least introduce some way to detect if non-sorted data is being added and panic.
